### ISSUE TITLE
Midi Mapping - Enhancement Request

### ISSUE THREAD

**Issue by malcolmjay** [monocasual/giada#384]
Although I have no issues with using the mouse most of the time, I think a few additions to the midi input actions would help the workflow using a just midi controller. Just a few ideas below:

-Making the start and end point of a sample in the sample editor window mappable 

-Have a feature where you can have up and down arrows to select a different channel and have them midi mappable. It would be nice to have this option too with just the keyboard arrow keys or through pageup/pagedown 

-Create a mappable button that brings up the option menu for a selected channel. From there a user could use the arrow buttons referenced above and have another mappable button to mimic a mouse click selection. You could have a similar option as well for the playback mode of the channel

**Comment by tomek-szczesny** [comment]
I'm certainly glad to see a request that seems to be very similar to my views regarding midi support in Giada! :) 

A few days ago I jumped into MIDI refactoring plan (see #375 and [wiki page](https://github.com/monocasual/giada/wiki/MIDI-overhaul)), because I also see a vast potential in controlling Giada from MIDI devices, and current implementation is, shall we say, not fond of major changes.

So this is a perfect time to discuss possible enhancements, so they could be included in the master plan. Also, if you feel like getting your hands dirty, a help with coding would be really appreciated, because I might as well spend a year doing all that myself :D 
However, I think that a separate (wiki?) space for gathering all ideas will be necessary in near future so none will be forgotten in the process.

Anyway, back to your requests: 
1. How do you envision mapping start and end point of a sample with a midi controller? I assume you mean a knob or fader. They usually emit CC messages that have 7 bit resolution, 0-127. I think this is hardly enough for such a task.

2. #221 is on a roadmap, which I think is necessary before implementing navigation keys, but yes, once this is done I agree that these keys could be mapped to MIDI controller as well.
What's important in #221 is actually a logic under the hood - what exactly should happen when channel is selected, and how is that executed.

3. Personally I don't believe it is possible to get rid of mouse control completely, although many common tasks should be indeed midi-mappable. The point of midi controllers is to make tasks easier or quicker, and I don't see navigating menus with midi controller buttons even close to a few mouse clicks. Could you perhaps describe a few examples of what would you do with such a feature in live setting? Maybe there are some other ways to achieve this.

Long time ago I described in the forums my idea for midi controller "function keys", that would work pretty much like Ctrl or Alt keys on a keyboard. So, for example, pressing a button corresponding with some channel while holding "arm" button, would arm or disarm a given channel. Same for mute, solo, edit (Action Editor) and whatnot. Simply pressing a key without any function key would do the default thing, start or stop a sample.
Maybe this approach would suit your taste?

**Comment by malcolmjay** [comment]
@tomek-szczesny I'm definitely open to helping out with coding. I just need some time to get up to speed with C++ and will start contributing once I'm comfortable with it.

Here's a link to what I started building a few days ago that's driving a lot of my recommendations: [](https://www.instagram.com/p/CDW0rOngvXy/?igshid=1gjvvs0wprllg)

For the start and end points, I think using a knob with midi CC is the best approach. That's exactly how Akai does it with their q-links.

**Comment by tomek-szczesny** [comment]
Love your little rig! Here's [mine](https://www.facebook.com/photo.php?fbid=1192573047766566&set=gm.3047195598711773&type=3&theater).

> For the start and end points, I think using a knob with midi CC is the best approach. That's exactly how Akai does it with their q-links.

Hm, I thought that 128 steps isn't enough to make it smooth, but on the other hand, a typical 4 seconds long 120bpm loop could be trimmed with about 31ms precision. That's not bad at all. So yeah, I'm convinced now.

And I'd love all the help I can get with MIDI overhaul, so don't hesitate joining the conversation in #375 once you feel ready. I'm not skilled in C++ myself, but do lots of reading to catch up.

**Comment by gvnnz** [comment]
See #805 for point 2. Closing.
